packages/context/src/value-promise.ts

```diff
@@ -191,32 +191,57 @@ export function resolveList<T, V>(
 export function tryWithFinally<T>(
   action: () => ValueOrPromise<T>,
   finalAction: () => void,
+): ValueOrPromise<T> {
+  const errorAction = (err: unknown): never => {
+    throw err;
+  };
+  return tryCatchFinally(action, errorAction, finalAction);
+}
+
+/**
+ * Try to run an action that returns a promise or a value with error and final
+ * actions to mimic `try {} catch(err) {} finally {}` for a value or promise.
+ *
+ * @param action - A function that returns a promise or a value
+ * @param errorAction - A function to be called once the action
+ * is rejected (synchronously or asynchronously). It must throw an error.
+ * @param finalAction - A function to be called once the action
+ * is fulfilled or rejected (synchronously or asynchronously)
+ */
+export function tryCatchFinally<T>(
+  action: () => ValueOrPromise<T>,
+  errorAction: (err: unknown) => never = err => {
+    throw err;
+  },
```

Comment:
What if the user wants to handle the error in such way that the entire operation resolves the final promise with a value? For example, we may want to catch "not found error" and convert it to "undefined" result.